pyfr/solvers/navstokes/elements.py

```diff
@@ -16,15 +20,78 @@ def set_backend(self, backend, nscalupts):
                        visc_corr=visc_corr,
                        c=self.cfg.items_as('constants', float))
 
+        shock_capturing = self.cfg.get('solver', 'shock-capturing', 'none')
+        if shock_capturing not in {'artificial-viscosity', 'none'}:
+            raise ValueError('Invalid shock-capturing option')
+
+        if shock_capturing == 'artificial-viscosity':
+            # Allocate required scratch space for artificial viscosity
+            nupts, nfpts, neles = self.nupts, self.nfpts, self.neles
+            tags = {'align'}
+            avis_upts = backend.matrix((nupts, 1, neles), extent='avis_upts', tags=tags)
+
+            if nfpts >= nupts:
+                self._avis_fpts = backend.matrix((nfpts, 1, neles), extent='avis_fpts', tags=tags)
+                avis_upts_temp = backend.matrix((nupts, 1, neles), aliases=self._avis_fpts, tags=tags)
+            else:
+                avis_upts_temp = backend.matrix((nupts, 1, neles), extent='avis_fpts', tags=tags)
+                self._avis_fpts = backend.matrix((nfpts, 1, neles), aliases=avis_upts_temp, tags=tags)
+
+            backend.pointwise.register('pyfr.solvers.navstokes.kernels.entropy')
+            backend.pointwise.register('pyfr.solvers.navstokes.kernels.avis')
+
+            def artf_vis():
+                # Compute entropy and save to avis_upts
+                ent = backend.kernel('entropy', tplargs=tplargs, dims=[self.nupts, self.neles],
+                                     u=self.scal_upts_inb, s=avis_upts)
+
+                # Compute modal coefficients of entropy and save them to avis_upts_temp
+                rcpvdm = np.linalg.inv(self.basis.ubasis.vdm.T)
+                rcpvdm = self._be.const_matrix(rcpvdm, tags={'align'})
+                mul = backend.kernel('mul', rcpvdm, avis_upts, out=avis_upts_temp)
+
+                # Additional constants for element-wise artificial viscosity kernel
+                ubdegs = self.basis.ubasis.degrees
+                tplargs['c'].update(self.cfg.items_as('solver-artificial-viscosity', float))
+                tplargs.update(dict(nupts=self.nupts, nfpts=self.nfpts,
+                                    order=self.basis.order, ubdegs=ubdegs))
+
+                # Column view for avis_upts/fpts matrices
+                def col_view(mat):
+                    vshape = (mat.nrow,)
+                    rcmap = np.array(list(ndrange(1, mat.ncol)))
+                    matmap = np.array([mat.mid]*mat.ncol)
+                    stridemap = np.array([[mat.leaddim]]*mat.ncol)
+
+                    return backend.view(matmap, rcmap, stridemap, vshape)
```

Comment:
Could you explain this a little bit more ? Do you mean that rcmap, matmap and stridmap, which is common, are pulled out from the col_view function ?